Check extension for latest version when executed

andyfeller opened the pull request:
<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->
Fixes #8183 

This PR applies the same approach used for checking and informing users of new `gh` releases to extensions when they are executed.  The benefit being that GitHub CLI users can be notified about new versions as they use the extension without putting that effort onto GitHub CLI extension authors.

> [!NOTE]
> This is initially created as a draft for feedback from core maintainers and the community before investing more effort to polish and complete the work.

### Demo

```shell
andrewfeller@Andrews-MacBook-Pro cli % gh ext install dlvhdr/gh-dash --pin v4.6.0
✓ Installed extension dlvhdr/gh-dash
✓ Pinned extension at v4.6.0
```

```shell
andrewfeller@Andrews-MacBook-Pro cli % ./bin/gh dash

A new release of dash is available: 4.6.0 → 4.7.0
To upgrade, run: gh extension upgrade dash --force
https://github.com/dlvhdr/gh-dash
```

### Open questions

1. **How frequently do we check for new extension releases?**

   The `gh` version checking only happens at minimum every 24 hours, relying upon information in XDG state directory on when it was last checked.  If the similar approach is desired here, then either the existing state file will be expanded OR `gh` maintains state information per extension.

   ```shell
   % cat ~/.local/state/gh/state.yml 
   checked_for_update_at: 2024-11-02T11:07:12.423561-04:00
   latest_release:
       version: v2.60.1
       url: https://github.com/cli/cli/releases/tag/v2.60.1
       publishedat: 2024-10-25T17:15:00Z
   ```

1. **What degree of testing is sufficient?**

   Being based on the existing behavior as Cobra command hooks, it is possible but want to solicit feedback before investing the time to understand any changes to the implementation.

andyfeller committed:
Prototype licensing workflow

andyfeller committed:
Fix working directory error

andyfeller committed:
Fix working directory error

andyfeller committed:
Fix go-licenses package targeting

andyfeller committed:
Fix Go root path issue

andyfeller committed:
Fix go-licenses template path error

andyfeller committed:
Fix go-licenses ordering

andyfeller committed:
Fix gh token usage

andyfeller committed:
Merge branch 'cli:trunk' into trunk

Commit:
Add version checking when executing extensions

Building on logic from the `gh ext list` for retrieving and assessing extension release information, this commit enhances the logic around invoking extensions to check for new releases.

Using the same user experience from checking `gh` version, this should only output information when the extension is used and gives the user information on how to upgrade depending on the type of extension and whether it is pinned or not.

```shell
andrewfeller@Andrews-MacBook-Pro cli % gh ext install dlvhdr/gh-dash --pin v4.6.0
✓ Installed extension dlvhdr/gh-dash
✓ Pinned extension at v4.6.0
andrewfeller@Andrews-MacBook-Pro cli % ./bin/gh dash

A new release of dash is available: 4.6.0 → 4.7.0
To upgrade, run: gh extension upgrade dash --force
https://github.com/dlvhdr/gh-dash
```

y-me-y commented on pkg/cmd/root/extension.go:
```diff
@@ -29,6 +54,24 @@ func NewCmdExtension(io *iostreams.IOStreams, em extensions.ExtensionManager, ex
+				stderr := io.ErrOut
+				fmt.Fprintf(stderr, "\n\n%s %s → %s\n",
+					ansi.Color(fmt.Sprintf("A new release of %s is available:", ext.Name()), "yellow"),
+					ansi.Color(strings.TrimPrefix(releaseInfo.CurrentVersion, "v"), "cyan"),
```
Total outside observer here but as a user I was wondering if it would make sense to differentiate color of current and latest versions. Make current more of orange or some warmer variant and keep the cool color tone for latest version.

andyfeller replied:
I'm open to the idea, let's talk about some considerations around [GitHub Primer Style](https://primer.style/foundations/color/overview#color-roles) and CLI accessibility.

Unlike web browsers, terminals have been designed to give users control over how [4-bit color](https://en.wikipedia.org/wiki/Color_depth#4-bit_color) is rendered when those color codes are used.  Unfortunately, this leaves limited options for a default color for those without color blindness:

1. black
1. red
1. green
1. yellow
1. blue
1. magenta
1. cyan
1. white
1. bright black
1. bright red
1. bright green
1. bright yellow
1. bright blue
1. bright magenta
1. bright cyan
1. bright white

![Screenshot 2024-11-04 at 3 56 27 PM](https://github.com/user-attachments/assets/2b0d34c4-791e-40e9-bda6-a75030b5f87b)

Building on that, the GitHub CLI has worked to apply GitHub's Primer Styles to our experience where applicable with one of those being [role of color](https://primer.style/foundations/color/overview#color-roles).  In this case, we've relied upon the `attention` role for information like this.

jtmcg replied:
Currently Implemented Colors:

<img width="741" alt="Screenshot 2024-11-05 at 11 14 51 AM" src="https://github.com/user-attachments/assets/d873601e-76ad-4137-ba09-eee304a15f30">

jtmcg replied:
Where'd your color choices come from, @andyfeller?

Also, should we wrap `gh extension upgrade dash --force` in `` ` ``?

andyfeller replied:
@jtmcg : They came from the `gh` version checking behavior below:

https://github.com/cli/cli/blob/215e4f440f56a0652e5c0f64614b64edcf1da1bc/internal/ghcmd/cmd.go#L169-L186

For now, I erred on the side of keeping this behavior 💯 consistent with `gh` version behavior as much as possible. 🤷

jtmcg replied:
For sure and I think that's a great choice 👍

jtmcg commented on pkg/cmd/root/extension.go:
```diff
@@ -4,20 +4,45 @@ import (
 
 	"github.com/cli/cli/v2/pkg/extensions"
 	"github.com/cli/cli/v2/pkg/iostreams"
+	"github.com/mgutz/ansi"
```
Why'd you choose to use this package instead of the color we have set up in the [iostreams package](https://github.com/cli/cli/blob/trunk/pkg/iostreams/color.go)?

andyfeller replied:
So, I originally copied the logic for checking `gh` version logic exactly and modified only what was needed:

https://github.com/cli/cli/blob/215e4f440f56a0652e5c0f64614b64edcf1da1bc/internal/ghcmd/cmd.go#L169-L186

That said, I did update this to rely upon the `ColorScheme` logic because it honors disabling colors when applicable 👍

jtmcg commented on pkg/cmd/root/extension.go:
```diff
@@ -4,20 +4,45 @@ import (
 		Short: fmt.Sprintf("Extension %s", ext.Name()),
+		// PreRun handles looking up whether extension has a latest version only when the command is ran.
+		PreRun: func(c *cobra.Command, args []string) {
+			go func() {
+				if ext.UpdateAvailable() {
+					updateMessageChan <- &extensionReleaseInfo{
+						CurrentVersion: ext.CurrentVersion(),
+						LatestVersion:  ext.LatestVersion(),
+						Pinned:         ext.IsPinned(),
+						URL:            ext.URL(),
+					}
+				}
+			}()
```
🤔 Something about this being an IIFE (instantly invoked functional expression) is giving me pause... That said, I'm not sure I've ever seen a goroutine created with a named function. Is that the convention?

Either way, I know you're interested in testing, and if this IIFE was extracted to a named function with its dependencies injected (`updateMessageChan`, `extensionReleaseInfo`, and `ext`) testing it would be pretty simple. It's a lot harder to test it here, as written.

andyfeller replied:
Just updated the PR with tests.  I don't think there's a lot of value testing channel handling behavior so much as if we're emitting the expected messages on standard error.

jtmcg replied:
Yeah, I agree... I didn't know the mocks existed for extensions already, so the tests for this are more straight forward than I anticipated ❤️ Definitely think you've nailed the tests

jtmcg reviewed (commented):
I like it! I'd ship the version-check as-is, I think, but you were looking for feedback so I've left a couple of questions.

Additionally, it's not clear to me what the inclusion of `licenses.tmpl` and `licenses.yml` files are for, here. Can you explain them to me?

andyfeller committed:
Test extension command update behaviors

This commit expands upon the previous work by creating tests around extension command execution and how various extension update scenarios are handled.

Along the way, the logic handling formatting update messaging has been switched to use `ColorScheme` in order to honor color behavior flags.

andyfeller commented:
After taking a second pass at these changes, I'll admit I'm second guessing how much we attempt to apply the same update caching logic used below around `gh`.  I'll explain my reasoning below but here is the code in question:

https://github.com/cli/cli/blob/215e4f440f56a0652e5c0f64614b64edcf1da1bc/internal/update/update.go#L22-L111

### Reasoning

The `gh` update checking logic is simple and straightforward:

1. `gh` has releases so we can make API calls to pull information
2. We're only checking a single GitHub repo for releases

Contrast that with extensions:

1. There are [3 types of extensions with very different mechanisms](https://github.com/cli/cli/blob/215e4f440f56a0652e5c0f64614b64edcf1da1bc/pkg/cmd/extension/extension.go#L88-L221) around version handling
   1. Local extension
   2. Remote script-based extension
   3. Remote binary-based extension
1. Local + remote binary-based extensions have [manifests](https://github.com/cli/cli/blob/215e4f440f56a0652e5c0f64614b64edcf1da1bc/pkg/cmd/extension/manager.go#L203-L211) similar to `gh` update state information however not exactly the same
1. These manifests are are the domain of the [extension manager](https://github.com/cli/cli/blob/215e4f440f56a0652e5c0f64614b64edcf1da1bc/pkg/cmd/extension/manager.go#L323-L335), possibly requiring a major refactor if we want to cache information per extension

With all of that said, I think there's enough done on the PR to bring it out of draft for further feedback.

andyfeller commented:
> I like it! I'd ship the version-check as-is, I think, but you were looking for feedback so I've left a couple of questions.
> 
> Additionally, it's not clear to me what the inclusion of `licenses.tmpl` and `licenses.yml` files are for, here. Can you explain them to me?

I'll be honest: I have no idea how that landed here.

andyfeller committed:
Delete .github/licenses.tmpl

andyfeller committed:
Delete .github/workflows/licenses.yml

jtmcg commented on pkg/cmd/root/extension_test.go:
```diff
@@ -0,0 +1,146 @@
+		extLatestVersion   string
+		extName            string
+		extUpdateAvailable bool
+		extURL             string
```
This doesn't seem to be used in any of the test cases. Should we remove it?

andyfeller replied:
Great catch, that's totally an oversight on my part and we should probably expand the `wantStderr` in the various test cases to check the url!

andyfeller replied:
Fixed!

jtmcg reviewed (approved):
Looks good to me! I've a tiny comment, but it's non-blocking 🚀

andyfeller committed:
Enhance extension upgrade tests for URL
